### ISSUE TITLE
Excessive memory usage when calling PutObject

### ISSUE THREAD

**Issue by exabytes18** [awslabs/aws-sdk-rust#623]
### Describe the bug

We have observed high memory usage when uploading large files to S3. The problem is compounded when performing multiple concurrent uploads and also when increasing the buffer size (as needed to achieve desired level of throughput to S3). Particularly, we need as many as 25 parallel uploads, which can result in several gigabyte of memory usage due to aws-sdk-rust alone. This is unexpected.

### Expected Behavior

Memory usage should be ~20MB for a single PutObject call.

### Current Behavior

Memory usage can reach 500MiB for a single PutObject call.

### Reproduction Steps

main.rs
```rust
extern crate bytesize;

use std::time::Duration;

use anyhow::Result;
use aws_sdk_s3::{Client, Region};
use aws_sdk_s3::types::ByteStream;
use bytesize::ByteSize;
use sysinfo::{get_current_pid, ProcessExt, System, SystemExt};
use tokio::time;

#[tokio::main]
async fn main() -> Result<()> {
    let pid = get_current_pid().unwrap();
    tokio::spawn(async move {
        let mut system = System::new();

        let mut interval = time::interval(Duration::from_secs(1));
        loop {
            interval.tick().await;
            system.refresh_process(pid);
            if let Some(process) = system.process(pid) {
                let memory = process.memory();
                println!("Memory usage: {}", ByteSize::b(memory).to_string());
            }
        }
    });

    println!("Sleeping for a few seconds before starting upload");
    time::sleep(Duration::from_secs(3)).await;

    let sdk_config = aws_config::from_env()
        .region(Region::new("us-east-1"))
        .load()
        .await;

    let s3_client = Client::new(&sdk_config);

    println!("Starting upload");

    s3_client.put_object()
        .bucket("my-bucket")
        .key("test")
        .body(ByteStream::read_from()
            .path("1gb.txt")
            .buffer_size(128 * 1024)
            .build()
            .await?)
        .send()
        .await?;

    Ok(())
}
```

cargo.toml
```rust
[package]
name = "s3-put-object-test"
version = "0.1.0"
edition = "2021"

[dependencies]
anyhow = "1.0"
aws-config = "0.49"
aws-sdk-s3 = "0.19"
bytesize = "1.1"
sysinfo = "0.26"
tokio = { version = "1", features = ["full"] }
```

Output (Local machine, home internet)
```
Sleeping for a few seconds before starting upload
Memory usage: 4.5 MB
Memory usage: 4.7 MB
Memory usage: 4.7 MB
Memory usage: 6.1 MB
Starting upload
Memory usage: 18.4 MB
Memory usage: 73.0 MB
Memory usage: 98.6 MB
Memory usage: 117.9 MB
Memory usage: 134.1 MB
Memory usage: 147.9 MB
Memory usage: 159.9 MB
Memory usage: 170.4 MB
Memory usage: 179.2 MB
Memory usage: 187.9 MB
Memory usage: 193.9 MB
Memory usage: 197.9 MB
Memory usage: 199.7 MB
Memory usage: 200.2 MB
Memory usage: 200.4 MB
Memory usage: 200.4 MB
Memory usage: 200.5 MB
Memory usage: 200.5 MB
Memory usage: 200.5 MB
Memory usage: 200.6 MB
Memory usage: 200.6 MB
Memory usage: 201.2 MB
Memory usage: 137.8 MB
Memory usage: 132.8 MB
Memory usage: 144.9 MB
Memory usage: 145.9 MB
Memory usage: 146.8 MB
Memory usage: 146.8 MB
Memory usage: 138.4 MB
Memory usage: 130.1 MB
Memory usage: 121.7 MB
Memory usage: 121.7 MB
Memory usage: 121.7 MB
Memory usage: 121.7 MB
Memory usage: 124.0 MB
Memory usage: 125.0 MB
Memory usage: 125.1 MB
Memory usage: 125.1 MB
Memory usage: 116.8 MB
Memory usage: 116.8 MB
Memory usage: 116.8 MB
Memory usage: 117.0 MB
Memory usage: 118.8 MB
Memory usage: 118.9 MB
Memory usage: 118.9 MB
Memory usage: 118.9 MB
Memory usage: 119.7 MB
Memory usage: 119.7 MB
Memory usage: 119.7 MB
Memory usage: 119.7 MB
Memory usage: 119.7 MB
Memory usage: 119.7 MB
Memory usage: 119.7 MB
Memory usage: 119.7 MB
Memory usage: 119.9 MB
Memory usage: 119.9 MB
```

Output (c6i.xlarge, repro app compiled with --release, cold page cache)
```
$ ./target/release/s3-put-object-test 
Sleeping for a few seconds before starting upload
Memory usage: 4.5 MB
Memory usage: 4.5 MB
Memory usage: 4.5 MB
Memory usage: 4.5 MB
Starting upload
Memory usage: 104.7 MB
Memory usage: 184.0 MB
Memory usage: 202.1 MB
Memory usage: 216.7 MB
Memory usage: 226.5 MB
Memory usage: 254.5 MB
Memory usage: 263.9 MB
Memory usage: 243.9 MB
Memory usage: 243.9 MB
```

Output (c6i.xlarge, repro app compiled with --release, warm page cache)
```
$ ./target/release/s3-put-object-test 
Sleeping for a few seconds before starting upload
Memory usage: 4.8 MB
Memory usage: 4.8 MB
Memory usage: 4.8 MB
Memory usage: 4.8 MB
Starting upload
Memory usage: 265.5 MB
Memory usage: 375.5 MB
Memory usage: 499.4 MB
Memory usage: 465.9 MB
Memory usage: 348.7 MB
```

### Version

```text
├── aws-config v0.49.0
│   ├── aws-http v0.49.0
│   │   ├── aws-smithy-http v0.49.0
│   │   │   ├── aws-smithy-eventstream v0.49.0
│   │   │   │   ├── aws-smithy-types v0.49.0
│   │   │   ├── aws-smithy-types v0.49.0 (*)
│   │   ├── aws-smithy-types v0.49.0 (*)
│   │   ├── aws-types v0.49.0
│   │   │   ├── aws-smithy-async v0.49.0
│   │   │   ├── aws-smithy-client v0.49.0
│   │   │   │   ├── aws-smithy-async v0.49.0 (*)
│   │   │   │   ├── aws-smithy-http v0.49.0 (*)
│   │   │   │   ├── aws-smithy-http-tower v0.49.0
│   │   │   │   │   ├── aws-smithy-http v0.49.0 (*)
│   │   │   │   ├── aws-smithy-types v0.49.0 (*)
│   │   │   ├── aws-smithy-http v0.49.0 (*)
│   │   │   ├── aws-smithy-types v0.49.0 (*)
│   ├── aws-sdk-sso v0.19.0
│   │   ├── aws-endpoint v0.49.0
│   │   │   ├── aws-smithy-http v0.49.0 (*)
│   │   │   ├── aws-smithy-types v0.49.0 (*)
│   │   │   ├── aws-types v0.49.0 (*)
│   │   ├── aws-http v0.49.0 (*)
│   │   ├── aws-sig-auth v0.49.0
│   │   │   ├── aws-sigv4 v0.49.0
│   │   │   │   ├── aws-smithy-eventstream v0.49.0 (*)
│   │   │   │   ├── aws-smithy-http v0.49.0 (*)
│   │   │   ├── aws-smithy-eventstream v0.49.0 (*)
│   │   │   ├── aws-smithy-http v0.49.0 (*)
│   │   │   ├── aws-types v0.49.0 (*)
│   │   ├── aws-smithy-async v0.49.0 (*)
│   │   ├── aws-smithy-client v0.49.0 (*)
│   │   ├── aws-smithy-http v0.49.0 (*)
│   │   ├── aws-smithy-http-tower v0.49.0 (*)
│   │   ├── aws-smithy-json v0.49.0
│   │   │   └── aws-smithy-types v0.49.0 (*)
│   │   ├── aws-smithy-types v0.49.0 (*)
│   │   ├── aws-types v0.49.0 (*)
│   ├── aws-sdk-sts v0.19.0
│   │   ├── aws-endpoint v0.49.0 (*)
│   │   ├── aws-http v0.49.0 (*)
│   │   ├── aws-sig-auth v0.49.0 (*)
│   │   ├── aws-smithy-async v0.49.0 (*)
│   │   ├── aws-smithy-client v0.49.0 (*)
│   │   ├── aws-smithy-http v0.49.0 (*)
│   │   ├── aws-smithy-http-tower v0.49.0 (*)
│   │   ├── aws-smithy-query v0.49.0
│   │   │   ├── aws-smithy-types v0.49.0 (*)
│   │   ├── aws-smithy-types v0.49.0 (*)
│   │   ├── aws-smithy-xml v0.49.0
│   │   ├── aws-types v0.49.0 (*)
│   ├── aws-smithy-async v0.49.0 (*)
│   ├── aws-smithy-client v0.49.0 (*)
│   ├── aws-smithy-http v0.49.0 (*)
│   ├── aws-smithy-http-tower v0.49.0 (*)
│   ├── aws-smithy-json v0.49.0 (*)
│   ├── aws-smithy-types v0.49.0 (*)
│   ├── aws-types v0.49.0 (*)
├── aws-sdk-s3 v0.19.0
│   ├── aws-endpoint v0.49.0 (*)
│   ├── aws-http v0.49.0 (*)
│   ├── aws-sig-auth v0.49.0 (*)
│   ├── aws-sigv4 v0.49.0 (*)
│   ├── aws-smithy-async v0.49.0 (*)
│   ├── aws-smithy-checksums v0.49.0
│   │   ├── aws-smithy-http v0.49.0 (*)
│   │   ├── aws-smithy-types v0.49.0 (*)
│   ├── aws-smithy-client v0.49.0 (*)
│   ├── aws-smithy-eventstream v0.49.0 (*)
│   ├── aws-smithy-http v0.49.0 (*)
│   ├── aws-smithy-http-tower v0.49.0 (*)
│   ├── aws-smithy-types v0.49.0 (*)
│   ├── aws-smithy-xml v0.49.0 (*)
│   ├── aws-types v0.49.0 (*)
```

### Environment details (OS name and version, etc.)

MacOS 12.6 (Monterey); Also observed on linux servers

**Comment by exabytes18** [comment]
I believe this might have been reported before, but it was closed without a solid repro case: https://github.com/awslabs/aws-sdk-rust/issues/453

Notably, I don't think the whole file is being buffered into memory, so the previous ticket might have a misleading title; however, memory usage does seem to be a good fraction of the file size (upwards of 50% of the file, in my tests).

The suggested comparison of using reqwest and using a presigned-url DOES help in our case, too. That is, I'm able to see high-throughput with only ~20MB memory usage; so, this high-memory usage is truly excessive.

**Comment by jdisanti** [comment]
Thank you for putting together this repro! This is immensely helpful!

I'm observing the same behavior locally. We will investigate.

**Comment by jdisanti** [comment]
I thoroughly read through all of the SDK's streaming code, and I don't see anything obvious in there.

I experimented with the repro, and it looks like the speed at which it consumes memory depends on the `ByteStream`'s `buffer_size`. A larger buffer size gets the memory use up to the full file size much faster.

I also experimented with letting the upload complete (with a 100 MB file) to see how memory use is afterwards. It looks like the memory use doesn't return back to pre-upload levels:

```
Sleeping for a few seconds before starting upload
Memory usage: 6.2 MB
Memory usage: 6.4 MB
Memory usage: 6.4 MB
Memory usage: 6.4 MB
Starting upload
Memory usage: 80.2 MB
Memory usage: 126.3 MB
Memory usage: 135.2 MB
...
Memory usage: 111.1 MB
Memory usage: 102.7 MB
Memory usage: 94.3 MB
Memory usage: 94.3 MB
Memory usage: 94.3 MB
Memory usage: 86.0 MB
...
Memory usage: 86.0 MB
Memory usage: 86.0 MB
Memory usage: 77.6 MB
Memory usage: 77.6 MB
Memory usage: 69.2 MB
Sleeping for a few seconds after finishing upload
Memory usage: 69.3 MB
Memory usage: 69.3 MB
Memory usage: 69.3 MB
```

Continuing to dig in.

**Comment by exabytes18** [comment]
Depending on the memory allocator behavior, the usage may not return to pre-upload levels (even if all memory is properly free'd by the application code and there is no leak). Can't speak to if there's any leak though.

**Comment by jdisanti** [comment]
I tried uploading to a presigned URL with reqwest as you described, and that definitely appears to use constant memory. Then I tried uploading to the presigned URL with hyper and hyper-rustls without any SDK logic, and that's also constant memory.

So then, that seems to imply the problem is with `ByteStream`, right? Since that's the variable eliminated here. So I modified my hyper presigned URL upload code to use `ByteStream`, and that also runs with constant memory:

```rust
#[tokio::main]
async fn main() -> Result<()> {
    let pid = get_current_pid().unwrap();
    tokio::spawn(async move {
        let mut system = System::new();

        let mut interval = time::interval(Duration::from_secs(1));
        loop {
            interval.tick().await;
            system.refresh_process(pid);
            if let Some(process) = system.process(pid) {
                let memory = process.memory();
                println!("Memory usage: {}", ByteSize::b(memory).to_string());
            }
        }
    });

    println!("Sleeping for a few seconds before starting upload");
    time::sleep(Duration::from_secs(3)).await;

    let sdk_config = aws_config::from_env()
        .region(Region::new("us-east-1"))
        .load()
        .await;

    let s3_client = Client::new(&sdk_config);

    let presigned = s3_client
        .put_object()
        .bucket("my-bucket")
        .key("100mb.txt")
        .content_length(100 * 1024 * 1024)
        .presigned(
            PresigningConfig::builder()
                .expires_in(Duration::from_secs(60))
                .build()
                .unwrap(),
        )
        .await?;

    println!("Starting upload");

    let client = hyper::client::Client::builder().build(
        hyper_rustls::HttpsConnectorBuilder::new()
            .with_native_roots()
            .https_only()
            .enable_http1()
            .build(),
    );

    let request = presigned
        .to_http_request(
            ByteStream::read_from()
                .path("100mb.txt")
                .buffer_size(128 * 1024)
                .build()
                .await
                .unwrap()
                .into_inner(),
        )
        .unwrap();
    let _response = client.request(request).await.unwrap();

    Ok(())
}
```

```
Sleeping for a few seconds before starting upload
Memory usage: 6.9 MB
Memory usage: 7.1 MB
Memory usage: 7.2 MB
Memory usage: 7.3 MB
Starting upload
Memory usage: 26.5 MB
Memory usage: 26.7 MB
Memory usage: 26.7 MB
Memory usage: 26.7 MB
Memory usage: 26.7 MB
Memory usage: 26.7 MB
Memory usage: 26.7 MB
Memory usage: 26.7 MB
Memory usage: 26.7 MB
Memory usage: 26.7 MB
Memory usage: 26.7 MB
Memory usage: 26.7 MB
Memory usage: 26.7 MB
Memory usage: 26.7 MB
Memory usage: 26.7 MB
Memory usage: 26.7 MB
Memory usage: 26.7 MB
Memory usage: 26.7 MB
Memory usage: 26.7 MB
Memory usage: 26.7 MB
Memory usage: 26.7 MB
Memory usage: 26.7 MB
Memory usage: 26.7 MB
Memory usage: 26.7 MB
Memory usage: 26.7 MB
Memory usage: 26.8 MB
...
```

This tells me it can't be `ByteStream`, which means it must be the default connector in the SDK. Or at least, that's the next thing I suspect.

**Comment by jdisanti** [comment]
It definitely seems to be the default connector, since replacing it with a vanilla hyper-rustls connector results in constant memory use:
```rust
let sdk_config = aws_config::from_env()
        .region(Region::new("us-east-2"))
        .load()
        .await;

    let connector = aws_smithy_client::hyper_ext::Adapter::builder().build(
        hyper_rustls::HttpsConnectorBuilder::new()
            .with_native_roots()
            .https_only()
            .enable_http1()
            .build(),
    );
    let s3_client = Client::from_conf_conn((&sdk_config).into(), connector);

    println!("Starting upload");
    s3_client
        .put_object()
        .bucket("aws-rust-sdk")
        .key("100mb.txt")
        .body(
            ByteStream::read_from()
                .path("100mb.txt")
                .buffer_size(128 * 1024)
                .build()
                .await
                .unwrap(),
        )
        .send()
        .await?;
```

```
Sleeping for a few seconds before starting upload
Memory usage: 7.1 MB
Memory usage: 7.3 MB
Memory usage: 7.3 MB
Memory usage: 7.5 MB
Starting upload
Memory usage: 22.6 MB
Memory usage: 22.7 MB
Memory usage: 22.7 MB
Memory usage: 22.8 MB
Memory usage: 22.9 MB
Memory usage: 22.9 MB
Memory usage: 22.9 MB
Memory usage: 23.0 MB
Memory usage: 23.0 MB
Memory usage: 23.0 MB
Memory usage: 23.0 MB
Memory usage: 23.2 MB
Memory usage: 23.3 MB
Memory usage: 23.3 MB
Memory usage: 23.4 MB
```
Looking into why the default connector has this behavior.

**Comment by jdisanti** [comment]
I think I found the issue. The current default connector is using the older 0.22 version of `hyper-rustls`, whereas the example above with constant memory use is using the latest 0.23 `hyper-rustls`. I modified the above example to use the older version, and immediately saw the unbounded memory growth.

So this issue will actually be resolved once [the timeout changes I've been working on](https://github.com/awslabs/smithy-rs/pull/1740) land since I've upgraded `hyper-rustls` in that PR.

Thanks again for reporting this! Hope to get the fix out soon.
